### ISSUE TITLE
AMP not used for training_step_end()

### ISSUE THREAD

**Issue by collinmccarthy** [Lightning-AI/lightning#8387]
## 🐛 Bug

It's possible this is intended or necessary, but the `torch.cuda.amp.autoscale()` context is used for, e.g. `training_step()` [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/f3e828426a41efa279c00ecdd8a86c96a79f6daa/pytorch_lightning/accelerators/accelerator.py#L199), but not for, e.g. `training_step_end()` [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/f3e828426a41efa279c00ecdd8a86c96a79f6daa/pytorch_lightning/accelerators/accelerator.py#L256).

### Expected behavior

If I want my code to work for both DDP and DP, I need to use `training_step_end()` so I can aggregate the results before computing my loss function, in this case `nll_loss`. As noted in the PTL docs this is especially important for something like `SimCLR` with a contrastive loss that needs many negative examples. And `nll_loss` is very unstable for 16-bit precision only (see [this issue here](https://github.com/pytorch/pytorch/issues/61523)). 

So people in this scenario will need `amp` for `training_step_end()` / `validation_step_end()` / `testing_step_end()`, and unless I'm missing something it makes sense to me that PTL does it automatically and not rely on the user to do so manually.

#### Thank you!

-Collin

**Comment by collinmccarthy** [comment]
I can submit a PR if the PTL team wants to call this in `training_step_end()` / `validation_step_end()` / `testing_step_end()` methods, just let me know. It looks to be as simple as mirroring the calls in the `_step()` methods since this is already the base `Accelerator` class. But please let me know if I'm wrong about that too.

**Comment by awaelchli** [comment]
@collinmccarthy if you can send a draft PR with your proposed solution, that would be great and I will be happy to take a look :) 

I don't think we call step_end() methods in DDP though. Maybe we should. If there is an edge case why we can't do that, I forgot about it but can't think of a reason right now.

**Comment by awaelchli** [comment]
Support for DataParallel (dp) got removed in #16748 for Lightning 2.0. Since this issue is very old, and was never picked up, I don't think it is worth including that in 1.9.x. 
If someone depends on this use case, let me know and we can brainstorm alternatives or ways of bringing a fix to 1.9.x
